src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceValueSerializer.cs

```diff
@@ -0,0 +1,128 @@
+            : base(ODataPayloadKind.Resource, serializerProvider)
+        {
+            if (serializerProvider == null)
+            {
+                throw Error.ArgumentNull("serializerProvider");
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ODataEdmTypeSerializer"/> class.
+        /// </summary>
+        /// <param name="payloadKind">The kind of OData payload that this serializer generates.</param>
+        /// <param name="serializerProvider">The <see cref="ODataSerializerProvider"/> to use to write inner objects.</param>
+        protected ODataResourceValueSerializer(ODataPayloadKind payloadKind, ODataSerializerProvider serializerProvider)
+            : base(payloadKind, serializerProvider)
+        {
+            if (serializerProvider == null)
+            {
+                throw Error.ArgumentNull("serializerProvider");
+            }
+          
+        }
+
+        /// <inheritdoc/>
+        public override void WriteObject(object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
+        {
+            if (messageWriter == null)
+            {
+                throw Error.ArgumentNull("messageWriter");
+            }
+            if (writeContext == null)
+            {
+                throw Error.ArgumentNull("writeContext");
+            }
+            if (writeContext.RootElementName == null)
+            {
+                throw Error.Argument("writeContext", SRResources.RootElementNameMissing, typeof(ODataSerializerContext).Name);
+            }
+
+            IEdmTypeReference edmType = writeContext.GetEdmType(graph, type);
+            Contract.Assert(edmType != null);
+
+            messageWriter.WriteProperty(CreateProperty(graph, edmType, writeContext.RootElementName, writeContext));
+        }
+
+        /// <inheritdoc/>
+        public sealed override ODataValue CreateODataValue(object graph, IEdmTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            if (!expectedType.IsComplex())
+            {
+                throw Error.InvalidOperation(SRResources.CannotWriteType, typeof(ODataResourceValueSerializer), expectedType.FullName());
+            }
+
+            ODataResourceValue value = CreateODataComplexResourceValue(graph, expectedType.AsComplex(), writeContext);
+            if (value == null)
+            {
+                return new ODataNullValue();
+            }
+
+            return value;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "edmTypeSerializer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        private ODataResourceValue CreateODataComplexResourceValue(object graph, IEdmComplexTypeReference expectedType, ODataSerializerContext writeContext)
```

Comment:
>CreateODataComplexResourceValue [](start = 35, length = 31)

Should this now be CreateODataResourceValue, if this is supposed to work for any resource?